File: frontend/src/types/exploration.ts
/**
 * Types for exploration feature.
 *
 * These types mirror the backend API schemas from spec 024 and add
 * frontend-specific types for tree visualization.
 */

import { z } from 'zod';

// =============================================================================
// Core Types (API Response)
// =============================================================================

/**
 * Goal definition for exploration.
 */
export interface ExplorationGoal {
  metric: 'success_rate';
  operator: '>=';
  value: number; // 0-1
}

/**
 * Configuration for exploration search.
 */
export interface ExplorationConfig {
  beam_width: number;
  max_depth: number;
  max_llm_calls: number;
  n_executions: number;
  sigma: number;
  seed: number | null;
}

/**
 * Status of an exploration.
 */
export type ExplorationStatus =
  | 'running'
  | 'goal_achieved'
  | 'depth_limit_reached'
  | 'cost_limit_reached'
  | 'no_viable_paths';

/**
 * Exploration entity from API.
 */
export interface Exploration {
  id: string;
  experiment_id: string;
  baseline_analysis_id: string;
  goal: ExplorationGoal;
  config: ExplorationConfig;
  status: ExplorationStatus;
  current_depth: number;
  total_nodes: number;
  total_llm_calls: number;
  best_success_rate: number | null;
  started_at: string;
  completed_at: string | null;
}

/**
 * Scorecard parameters for a scenario.
 */
export interface ScorecardParams {
  complexity: number;
  initial_effort: number;
  perceived_risk: number;
  time_to_value: number;
}

/**
 * Simulation results for a scenario.
 */
export interface SimulationResults {
  success_rate: number;
  fail_rate: number;
  did_not_try_rate: number;
}

/**
 * Status of a scenario node.
 */
export type NodeStatus = 'active' | 'dominated' | 'winner' | 'expansion_failed';

/**
 * Scenario node from API.
 */
export interface ScenarioNode {
  id: string;
  exploration_id: string;
  parent_id: string | null;
  depth: number;
  action_applied: string | null;
  action_category: string | null;
  rationale: string | null;
  scorecard_params: ScorecardParams;
  simulation_results: SimulationResults | null;
  execution_time_seconds: number | null;
  node_status: NodeStatus;
  created_at: string;
}

/**
 * Exploration tree response from API.
 */
export interface ExplorationTree {
  exploration: Exploration;
  nodes: ScenarioNode[];
  node_count_by_status: Record<NodeStatus, number>;
}

/**
 * Path step in winning path.
 */
export interface PathStep {
  depth: number;
  action: string | null;
  category: string | null;
  rationale: string | null;
  success_rate: number;
  delta_success_rate: number;
}

/**
 * Winning path response from API.
 */
export interface WinningPath {
  exploration_id: string;
  winner_node_id: string;
  path: PathStep[];
  total_improvement: number;
}

// =============================================================================
// Action Catalog Types
// =============================================================================

/**
 * Impact range for an action parameter.
 */
export interface ImpactRange {
  min: number;
  max: number;
}

/**
 * Example action in catalog.
 */
export interface ActionExample {
  action: string;
  typical_impacts: Record<string, ImpactRange>;
}

/**
 * Action category in catalog.
 */
export interface ActionCategory {
  id: string;
  name: string;
  description: string;
  examples: ActionExample[];
}

/**
 * Action catalog response.
 */
export interface ActionCatalog {
  version: string;
  categories: ActionCategory[];
}

// =============================================================================
// Request Types
// =============================================================================

/**
 * Request to create a new exploration.
 */
export interface ExplorationCreate {
  experiment_id: string;
  goal_value: number;
  beam_width?: number;
  max_depth?: number;
  max_llm_calls?: number;
  n_executions?: number;
  seed?: number | null;
}

// =============================================================================
// Tree Visualization Types
// =============================================================================

/**
 * Node data for react-d3-tree.
 * Extends base node with visualization properties.
 */
export interface TreeNodeData {
  name: string; // Display label (success_rate %)
  attributes?: {
    action?: string; // Truncated action text
    status: NodeStatus; // For color coding
  };
  children?: TreeNodeData[];
  // Original node data for detail panel
  __nodeData: ScenarioNode;
}

/**
 * Props for custom node element.
 */
export interface CustomNodeProps {
  nodeDatum: TreeNodeData;
  toggleNode: () => void;
  onNodeClick?: (node: ScenarioNode) => void;
}

/**
 * Colors for node status.
 */
export const NODE_STATUS_COLORS: Record<NodeStatus, string> = {
  winner: '#22c55e', // green-500
  active: '#3b82f6', // blue-500
  dominated: '#94a3b8', // slate-400
  expansion_failed: '#ef4444', // red-500
};

/**
 * Badge variants for node status.
 */
export const NODE_STATUS_BADGES: Record<
  NodeStatus,
  {
    label: string;
    variant: 'success' | 'info' | 'neutral' | 'error';
  }
> = {
  winner: { label: 'Vencedor', variant: 'success' },
  active: { label: 'Ativo', variant: 'info' },
  dominated: { label: 'Dominado', variant: 'neutral' },
  expansion_failed: { label: 'Falhou', variant: 'error' },
};

// =============================================================================
// Exploration List Types (for experiment page)
// =============================================================================

/**
 * Summary of exploration for list view.
 */
export interface ExplorationSummary {
  id: string;
  status: ExplorationStatus;
  goal_value: number;
  best_success_rate: number | null;
  total_nodes: number;
  started_at: string;
  completed_at: string | null;
}

// =============================================================================
// Zod Schema (for form validation)
// =============================================================================

export const explorationCreateSchema = z.object({
  experiment_id: z.string().min(1, 'ID do experimento é obrigatório'),
  goal_value: z
    .number()
    .min(0.01, 'Meta deve ser maior que 0%')
    .max(1, 'Meta deve ser menor ou igual a 100%'),
  beam_width: z
    .number()
    .int()
    .min(1, 'Beam width mínimo é 1')
    .max(10, 'Beam width máximo é 10')
    .default(3),
  max_depth: z
    .number()
    .int()
    .min(1, 'Profundidade mínima é 1')
    .max(10, 'Profundidade máxima é 10')
    .default(5),
});

export type ExplorationCreateForm = z.infer<typeof explorationCreateSchema>;

File: src/synth_lab/api/schemas/exploration.py
"""
API schemas for exploration endpoints.

Request and response models for LLM-assisted scenario exploration.

References:
    - Spec: specs/024-llm-scenario-exploration/spec.md
    - API contract: specs/024-llm-scenario-exploration/contracts/exploration-api.yaml
"""

from datetime import datetime

from pydantic import BaseModel, Field

# ========== Request Schemas ==========


class ExplorationCreate(BaseModel):
    """Request to create a new exploration."""

    experiment_id: str = Field(
        pattern=r"^exp_[a-f0-9]{8}$",
        description="ID of the source experiment.",
        json_schema_extra={"example": "exp_12345678"})

    goal_value: float = Field(
        ge=0.0,
        le=1.0,
        description="Target success_rate to achieve (0-1).",
        json_schema_extra={"example": 0.40})

    beam_width: int = Field(
        default=3,
        ge=1,
        le=10,
        description="Number of scenarios to keep per iteration.")

    max_depth: int = Field(
        default=5,
        ge=1,
        le=10,
        description="Maximum depth of exploration tree.")

    max_llm_calls: int = Field(
        default=20,
        ge=1,
        le=100,
        description="Maximum number of LLM calls.")

    n_executions: int = Field(
        default=100,
        ge=10,
        le=1000,
        description="Monte Carlo executions per simulation.")

    seed: int | None = Field(
        default=None,
        description="Random seed for reproducibility.")


# ========== Response Schemas ==========


class GoalResponse(BaseModel):
    """Goal definition in response."""

    metric: str = Field(description="Metric to optimize (success_rate).")
    operator: str = Field(description="Comparison operator (>=).")
    value: float = Field(description="Target value.")


class ConfigResponse(BaseModel):
    """Configuration in response."""

    beam_width: int
    max_depth: int
    max_llm_calls: int
    n_executions: int
    sigma: float
    seed: int | None


class ExplorationSummary(BaseModel):
    """Summary of exploration for list view."""

    id: str = Field(description="Exploration ID.")
    status: str = Field(description="Current status.")
    goal_value: float = Field(description="Target success_rate.")
    best_success_rate: float | None = Field(description="Best success rate achieved.")
    total_nodes: int = Field(description="Total nodes created.")
    started_at: datetime = Field(description="Start timestamp.")
    completed_at: datetime | None = Field(description="Completion timestamp.")


class ExplorationResponse(BaseModel):
    """Response for exploration data."""

    id: str = Field(description="Exploration ID.")
    experiment_id: str = Field(description="Source experiment ID.")
    baseline_analysis_id: str = Field(description="Baseline analysis ID.")
    goal: GoalResponse = Field(description="Performance goal.")
    config: ConfigResponse = Field(description="Search configuration.")
    status: str = Field(description="Current status.")
    current_depth: int = Field(description="Current depth reached.")
    total_nodes: int = Field(description="Total nodes created.")
    total_llm_calls: int = Field(description="Total LLM calls made.")
    best_success_rate: float | None = Field(description="Best success rate achieved.")
    started_at: datetime = Field(description="Start timestamp.")
    completed_at: datetime | None = Field(description="Completion timestamp.")


class ScorecardParamsResponse(BaseModel):
    """Scorecard parameters in response."""

    complexity: float
    initial_effort: float
    perceived_risk: float
    time_to_value: float


class SimulationResultsResponse(BaseModel):
    """Simulation results in response."""

    success_rate: float
    fail_rate: float
    did_not_try_rate: float


class ScenarioNodeResponse(BaseModel):
    """Response for a scenario node."""

    id: str = Field(description="Node ID.")
    exploration_id: str = Field(description="Parent exploration ID.")
    parent_id: str | None = Field(description="Parent node ID.")
    depth: int = Field(description="Depth in tree.")
    action_applied: str | None = Field(description="Action that created this node.")
    action_category: str | None = Field(description="Category of action.")
    rationale: str | None = Field(description="Rationale for action.")
    short_action: str | None = Field(description="Short 3-word action label.")
    scorecard_params: ScorecardParamsResponse = Field(description="Scorecard parameters.")
    simulation_results: SimulationResultsResponse | None = Field(
        description="Simulation results."
    )
    execution_time_seconds: float | None = Field(description="Simulation time.")
    node_status: str = Field(description="Node status.")
    created_at: datetime = Field(description="Creation timestamp.")


class ExplorationTreeResponse(BaseModel):
    """Response for exploration tree."""

    exploration: ExplorationResponse = Field(description="Exploration data.")
    nodes: list[ScenarioNodeResponse] = Field(description="All nodes in tree.")
    node_count_by_status: dict[str, int] = Field(description="Count by status.")


class PathStepResponse(BaseModel):
    """A step in the winning path."""

    depth: int
    action: str | None
    category: str | None
    rationale: str | None
    success_rate: float
    delta_success_rate: float


class WinningPathResponse(BaseModel):
    """Response for winning path."""

    exploration_id: str = Field(description="Exploration ID.")
    winner_node_id: str = Field(description="Winner node ID.")
    path: list[PathStepResponse] = Field(description="Path from root to winner.")
    total_improvement: float = Field(description="Total success_rate improvement.")


class IterationResultResponse(BaseModel):
    """Response for iteration result."""

    exploration_id: str
    iteration_number: int
    status: str
    nodes_expanded: int
    nodes_created: int
    nodes_dominated: int
    llm_calls_made: int
    best_success_rate: float
    frontier_size: int


# ========== Catalog Schemas ==========


class ImpactRangeResponse(BaseModel):
    """Impact range for an action example."""

    min: float
    max: float


class ActionExampleResponse(BaseModel):
    """Example action in catalog."""

    action: str
    typical_impacts: dict[str, ImpactRangeResponse]


class ActionCategoryResponse(BaseModel):
    """Action category in catalog."""

    id: str
    name: str
    description: str
    examples: list[ActionExampleResponse]


class ActionCatalogResponse(BaseModel):
    """Response for action catalog."""

    version: str
    categories: list[ActionCategoryResponse]


# ========== Conversion Functions ==========


def exploration_to_response(exploration) -> ExplorationResponse:
    """Convert Exploration entity to response schema."""
    return ExplorationResponse(
        id=exploration.id,
        experiment_id=exploration.experiment_id,
        baseline_analysis_id=exploration.baseline_analysis_id,
        goal=GoalResponse(
            metric=exploration.goal.metric,
            operator=exploration.goal.operator,
            value=exploration.goal.value),
        config=ConfigResponse(
            beam_width=exploration.config.beam_width,
            max_depth=exploration.config.max_depth,
            max_llm_calls=exploration.config.max_llm_calls,
            n_executions=exploration.config.n_executions,
            sigma=exploration.config.sigma,
            seed=exploration.config.seed),
        status=exploration.status.value,
        current_depth=exploration.current_depth,
        total_nodes=exploration.total_nodes,
        total_llm_calls=exploration.total_llm_calls,
        best_success_rate=exploration.best_success_rate,
        started_at=exploration.started_at,
        completed_at=exploration.completed_at)


def node_to_response(node) -> ScenarioNodeResponse:
    """Convert ScenarioNode entity to response schema."""
    simulation_results = None
    if node.simulation_results:
        simulation_results = SimulationResultsResponse(
            success_rate=node.simulation_results.success_rate,
            fail_rate=node.simulation_results.fail_rate,
            did_not_try_rate=node.simulation_results.did_not_try_rate)

    return ScenarioNodeResponse(
        id=node.id,
        exploration_id=node.exploration_id,
        parent_id=node.parent_id,
        depth=node.depth,
        action_applied=node.action_applied,
        action_category=node.action_category,
        rationale=node.rationale,
        short_action=node.short_action,
        scorecard_params=ScorecardParamsResponse(
            complexity=node.scorecard_params.complexity,
            initial_effort=node.scorecard_params.initial_effort,
            perceived_risk=node.scorecard_params.perceived_risk,
            time_to_value=node.scorecard_params.time_to_value),
        simulation_results=simulation_results,
        execution_time_seconds=node.execution_time_seconds,
        node_status=node.node_status.value,
        created_at=node.created_at)
